File: branchbound.java
import java.util.Scanner;

public class branchbound {

    static int n;
    static int[][] cost;
    static boolean[] visited;
    static int[] bestPath;
    static int[] currPath;

    static int bestCost = Integer.MAX_VALUE;

    // Function to calculate lower bound (minimum edge cost sum approximation)
    static int calculateBound(int currentCost) {
        int bound = currentCost;

        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                int minEdge = Integer.MAX_VALUE;

                for (int j = 0; j < n; j++) {
                    if (i != j && cost[i][j] != 0) {
                        minEdge = Math.min(minEdge, cost[i][j]);
                    }
                }
                bound += minEdge;
            }
        }
        return bound;
    }

    // Branch and Bound Recursive Function
    static void tsp(int level, int currentCity, int currentCost) {

        // If all cities are visited, check returning to start
        if (level == n) {
            if (cost[currentCity][0] != 0) {
                int totalCost = currentCost + cost[currentCity][0];

                if (totalCost < bestCost) {
                    bestCost = totalCost;

                    for (int i = 0; i < n; i++)
                        bestPath[i] = currPath[i];

                    bestPath[n] = 0; // return to start
                }
            }
            return;
        }

        // Try all next cities
        for (int nextCity = 0; nextCity < n; nextCity++) {

            if (!visited[nextCity] && cost[currentCity][nextCity] != 0) {

                visited[nextCity] = true;
                currPath[level] = nextCity;

                int newCost = currentCost + cost[currentCity][nextCity];

                // Calculate Bound
                int bound = calculateBound(newCost);

                // Prune if bound is worse than best cost
                if (bound < bestCost) {
                    tsp(level + 1, nextCity, newCost);
                }

                visited[nextCity] = false;
            }
        }
    }

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of cities: ");
        n = sc.nextInt();

        cost = new int[n][n];
        visited = new boolean[n];
        bestPath = new int[n + 1];
        currPath = new int[n];

        System.out.println("Enter cost adjacency matrix (0 if no path): ");
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                cost[i][j] = sc.nextInt();
            }
        }

        // Start from city 0
        visited[0] = true;
        currPath[0] = 0;

        tsp(1, 0, 0);

        System.out.println("\n===== OUTPUT =====");
        if (bestCost == Integer.MAX_VALUE) {
            System.out.println("No solution exists (graph disconnected).");
        } else {
            System.out.println("Minimum Cost = " + bestCost);
            System.out.print("Best Path = ");

            for (int i = 0; i <= n; i++) {
                System.out.print(bestPath[i] + " ");
            }
            System.out.println();
        }

        sc.close();
    }
}

File: branchbound.c
#include <stdio.h>
#include <limits.h>

#define MAX 10

int n;
int cost[MAX][MAX];
int visited[MAX];
int finalPath[MAX];
int tempPath[MAX];

int finalCost = INT_MAX;

// Function to find minimum edge cost from a city
int firstMin(int i)
{
    int min = INT_MAX;
    for(int k = 0; k < n; k++)
    {
        if(cost[i][k] < min && i != k)
            min = cost[i][k];
    }
    return min;
}

// Function to find second minimum edge cost from a city
int secondMin(int i)
{
    int first = INT_MAX, second = INT_MAX;

    for(int j = 0; j < n; j++)
    {
        if(i == j)
            continue;

        if(cost[i][j] <= first)
        {
            second = first;
            first = cost[i][j];
        }
        else if(cost[i][j] < second && cost[i][j] != first)
        {
            second = cost[i][j];
        }
    }
    return second;
}

// Branch and Bound recursive function
void TSPRec(int currBound, int currWeight, int level, int currPath[])
{
    if(level == n)
    {
        if(cost[currPath[level - 1]][currPath[0]] != 0)
        {
            int currRes = currWeight + cost[currPath[level - 1]][currPath[0]];

            if(currRes < finalCost)
            {
                for(int i = 0; i < n; i++)
                    finalPath[i] = currPath[i];

                finalPath[n] = currPath[0];
                finalCost = currRes;
            }
        }
        return;
    }

    for(int i = 0; i < n; i++)
    {
        if(cost[currPath[level - 1]][i] != 0 && visited[i] == 0)
        {
            int tempBound = currBound;
            currWeight += cost[currPath[level - 1]][i];

            if(level == 1)
                currBound -= ((firstMin(currPath[level - 1]) + firstMin(i)) / 2);
            else
                currBound -= ((secondMin(currPath[level - 1]) + firstMin(i)) / 2);

            // If promising
            if(currBound + currWeight < finalCost)
            {
                currPath[level] = i;
                visited[i] = 1;
                TSPRec(currBound, currWeight, level + 1, currPath);
            }

            // Backtrack
            currWeight -= cost[currPath[level - 1]][i];
            currBound = tempBound;

            for(int j = 0; j < n; j++)
                visited[j] = 0;

            for(int j = 0; j < level; j++)
                visited[currPath[j]] = 1;
        }
    }
}

// Main TSP function
void TSP(int start)
{
    int currPath[MAX];

    int currBound = 0;
    for(int i = 0; i < n; i++)
        currBound += (firstMin(i) + secondMin(i));

    currBound = (currBound % 2) ? (currBound / 2 + 1) : (currBound / 2);

    for(int i = 0; i < n; i++)
        visited[i] = 0;

    visited[start] = 1;
    currPath[0] = start;

    TSPRec(currBound, 0, 1, currPath);
}

int main()
{
    int start;

    printf("Enter number of cities: ");
    scanf("%d", &n);

    printf("Enter cost adjacency matrix:\n");
    for(int i = 0; i < n; i++)
    {
        for(int j = 0; j < n; j++)
        {
            scanf("%d", &cost[i][j]);
        }
    }

    printf("Enter starting city (0 to %d): ", n-1);
    scanf("%d", &start);

    TSP(start);

    printf("\nMinimum Cost = %d\n", finalCost);
    printf("Optimal Path: ");
    for(int i = 0; i <= n; i++)
        printf("%d ", finalPath[i]);

    printf("\n");

    return 0;
}

